src/librustc/traits/on_unimplemented.rs

```diff
@@ -0,0 +1,296 @@
+
+        for command in self.subcommands.iter().chain(Some(self)).rev() {
+            if let Some(ref condition) = command.condition {
+                if !attr::eval_condition(condition, &tcx.sess.parse_sess, &mut |c| {
+                    options.iter().any(|o| c.check_name(o))
+                }) {
+                    debug!("evaluate: skipping {:?} due to condition", command);
+                    continue
+                }
+            }
+            debug!("evaluate: {:?} succeeded", command);
+            message = command.message.clone();
+            label = command.label.clone();
+        }
+
+        OnUnimplementedNote {
+            label: label.map(|l| l.format(tcx, trait_ref)),
+            message: message.map(|m| m.format(tcx, trait_ref))
+        }
+    }
+}
+
+impl<'a, 'gcx, 'tcx> OnUnimplementedFormatString {
+    pub fn try_parse(tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                     trait_def_id: DefId,
+                     from: InternedString,
+                     err_sp: Span)
+                     -> Result<Self, ErrorReported>
+    {
+        let result = OnUnimplementedFormatString(from);
+        result.verify(tcx, trait_def_id, err_sp)?;
+        Ok(result)
+    }
+
+    fn verify(&self,
+              tcx: TyCtxt<'a, 'gcx, 'tcx>,
+              trait_def_id: DefId,
+              span: Span)
+              -> Result<(), ErrorReported>
+    {
+        let name = tcx.item_name(trait_def_id).as_str();
+        let generics = tcx.generics_of(trait_def_id);
+        let parser = Parser::new(&self.0);
+        let types = &generics.types;
+        let mut result = Ok(());
+        for token in parser {
+            match token {
+                Piece::String(_) => (), // Normal string, no need to check it
+                Piece::NextArgument(a) => match a.position {
+                    // `{Self}` is allowed
+                    Position::ArgumentNamed(s) if s == "Self" => (),
+                    // `{ThisTraitsName}` is allowed
+                    Position::ArgumentNamed(s) if s == name => (),
+                    // So is `{A}` if A is a type parameter
+                    Position::ArgumentNamed(s) => match types.iter().find(|t| {
+                        t.name == s
+                    }) {
+                        Some(_) => (),
+                        None => {
+                            span_err!(tcx.sess, span, E0230,
+                                      "there is no type parameter \
+                                       {} on trait {}",
+                                      s, name);
+                            result = Err(ErrorReported);
+                        }
+                    },
+                    // `{:1}` and `{}` are not to be used
+                    Position::ArgumentIs(_) => {
+                        span_err!(tcx.sess, span, E0231,
+                                  "only named substitution \
+                                   parameters are allowed");
```

Comment:
Add a note including all the possible valid named parameters?

Reply:
That's preexisting.